src/mlang/backend_ir/bir.ml

```diff
@@ -27,32 +29,51 @@ and stmt_kind =
   | SConditional of Mir.expression * stmt list * stmt list
   | SVerif of Mir.condition_data
   | SRuleCall of rule_id
+  | SFunctionCall of function_name * Mir.Variable.t list
+
+type mpp_function = stmt list
+
+module FunctionMap = Map.Make (struct
+  type t = function_name
+
+  let compare = String.compare
+end)
 
 type program = {
+  mpp_functions : mpp_function FunctionMap.t;
   rules : rule RuleMap.t;
-  statements : stmt list;
+  main_function : function_name;
   idmap : Mir.idmap;
   mir_program : Mir.program;
   outputs : unit Mir.VariableMap.t;
 }
 
-let rec get_block_statements (rules : rule RuleMap.t) (stmts : stmt list) :
-    stmt list =
+let main_statements (p : program) : stmt list =
+  try FunctionMap.find p.main_function p.mpp_functions
+  with Not_found ->
+    Errors.raise_error "Unable to find main function of Bir program"
+
+let rec get_block_statements (rules : rule RuleMap.t) (p : program)
```

Comment:
Since you introduced `p`, could you remove `rules` which is already a field of `p` ?